### PR TITLE
Delete transferred files

### LINKED ISSUE #374: [Bug]: tar files are not deleted after successful globus transfer
### What happened?

The documentation indicates that "--keep to keep a copy of the tar files on the local file system after they have been transferred to HPSS. Normally, they are deleted after successful transfer." So, without `--keep`, the tar files should be deleted after successful transfer. But the tar files are not deleted when transferred successfully via globus. I might not use the zstash command correctly. If so, sorry for the false alarm.

I reproduced this with a simple test on Chrysalis.
`touch tst.txt`
`zstash create -v --non-blocking --hpss=globus://nersc/~/E3SMv3_dev/tst --maxsize 128 .`

The log line `Calling hpss_put to dispatch archive file 000000.tar [keep, non_blocking] = [False, True]` confirms the keep flag is `False`.

Here is the complete log:
```
For help, please see https://e3sm-project.github.io/zstash. Ask questions at https://github.com/E3SM-Project/zstash/discussions/categories/q-a.
DEBUG: 20250605_035238_623132: Running zstash create
DEBUG: Local path : /lcrc/group/e3sm2/ac.qtang/tst
DEBUG: HPSS path  : globus://nersc/~/E3SMv3_dev/tst
DEBUG: Max size  : 137438953472
DEBUG: Keep local tar files  : False
DEBUG: Making sure input path exists and is a directory
DEBUG: 20250605_035238_623404:Calling globus_activate(hpss)
INFO: Creating client of type <class 'globus_sdk.services.auth.client.native_client.NativeAppAuthClient'> for service "auth"
INFO: Finished initializing AuthLoginClient. client_id='6c1629cf-446c-49e7-af95-323c6412397f', type(authorizer)=<class 'globus_sdk.authorizers.base.NullAuthorizer'>
INFO: Setting up RefreshTokenAuthorizer with auth_client=[instance:23455989535200]
INFO: Setting up a RenewingAuthorizer. It will use an auth type of Bearer and can handle 401s.
INFO: RenewingAuthorizer will start by using access_token with hash "6e0977482bceb074128f911539d014c5907e30d53bcf7fa9c0c1d1b7240ff704"
INFO: Setting up an AccessTokenAuthorizer. It will use an auth type of Bearer and cannot handle 401s.
INFO: Creating client of type <class 'globus_sdk.services.transfer.client.TransferClient'> for service "transfer"
INFO: TransferClient.endpoint_autoactivate(15288284-7006-4041-ba1a-6b52501e49f1)
INFO: request done (success)
INFO: TransferClient.endpoint_autoactivate(9cd89cfd-6d04-11e5-ba46-22000b92c6ec)
INFO: request done (success)
DEBUG: 20250605_035239_319250: Creating local cache directory
DEBUG: 20250605_035239_319398: Calling create_database()
DEBUG: 20250605_035239_319454:Creating index database
INFO: Gathering list of files to archive
INFO: 20250605_035239_420308: Creating new tar archive 000000.tar
INFO: Archiving tst.txt
DEBUG: 20250605_035239_423862: Closing tar archive 000000.tar
INFO: 20250605_035239_424067: (add_files): Completed archive file 000000.tar
INFO: tar name=000000.tar, tar size=10240, tar md5=21d45155d348307f7db791b620a07d96
INFO: 20250605_035239_429147: DIVING: (add_files): Calling hpss_put to dispatch archive file 000000.tar [keep, non_blocking] = [False, True]
INFO: 20250605_035239_429204: in hpss_transfer, prev_transfers is starting as []
INFO: Transferring file to HPSS: zstash/000000.tar
INFO: 20250605_035239_429272: DIVING: hpss calls globus_transfer(name=000000.tar)
INFO: 20250605_035239_429301: Entered globus_transfer() for name = 000000.tar
DEBUG: 20250605_035239_429322: non_blocking = True
INFO: Creating a new TransferData object
INFO: TransferClient.get_submission_id(None)
INFO: request done (success)
INFO: TransferData.DATA_TYPE = transfer
INFO: TransferData.DATA = []
INFO: TransferData.source_endpoint = 15288284-7006-4041-ba1a-6b52501e49f1
INFO: TransferData.destination_endpoint = 9cd89cfd-6d04-11e5-ba46-22000b92c6ec
INFO: TransferData.label = tst 000000
INFO: TransferData.submission_id = 867de2ca-41c0-11f0-8ee3-0256b1e82d1f
INFO: TransferData.verify_checksum = True
INFO: TransferData.preserve_timestamp = True
INFO: TransferData.encrypt_data = False
INFO: TransferData.skip_source_errors = False
INFO: TransferData.fail_on_quota_errors = True
INFO: TransferData.delete_destination_extra = False
INFO: TransferData.notify_on_succeeded = True
INFO: TransferData.notify_on_failed = True
INFO: TransferData.notify_on_inactive = True
INFO: 20250605_035239_455518: TransferData: accumulated items:
   (routine)  PUSHING (#1) STORED source item: /lcrc/group/e3sm2/ac.qtang/tst/zstash/000000.tar
INFO: 20250605_035239_455593: DIVING: Submit Transfer for tst 000000
INFO: TransferClient.submit_transfer(...)
INFO: request done (success)
INFO: 20250605_035241_289602: SURFACE Submit Transfer returned new task_id = 867de1e9-41c0-11f0-b34a-0256b1e82d1f for label tst 000000
INFO: 20250605_035241_289676: NO BLOCKING (task_wait) for task_id 867de1e9-41c0-11f0-b34a-0256b1e82d1f
INFO: 20250605_035241_289736: SURFACE hpss globus_transfer(name=000000.tar) returns UNKNOWN
INFO: 20250605_035241_289815: SURFACE (add_files): Called hpss_put to dispatch archive file 000000.tar
DEBUG: 20250605_035241_296579: calling hpss_put() for zstash/index.db
INFO: 20250605_035241_296765: in hpss_transfer, prev_transfers is starting as []
INFO: Transferring file to HPSS: zstash/index.db
INFO: 20250605_035241_296862: DIVING: hpss calls globus_transfer(name=index.db)
INFO: 20250605_035241_296899: Entered globus_transfer() for name = index.db
DEBUG: 20250605_035241_296922: non_blocking = False
INFO: Creating a new TransferData object
INFO: TransferClient.get_submission_id(None)
INFO: request done (success)
INFO: TransferData.DATA_TYPE = transfer
INFO: TransferData.DATA = []
INFO: TransferData.source_endpoint = 15288284-7006-4041-ba1a-6b52501e49f1
INFO: TransferData.destination_endpoint = 9cd89cfd-6d04-11e5-ba46-22000b92c6ec
INFO: TransferData.label = tst index
INFO: TransferData.submission_id = 879ade90-41c0-11f0-aad5-0256b1e82d1f
INFO: TransferData.verify_checksum = True
INFO: TransferData.preserve_timestamp = True
INFO: TransferData.encrypt_data = False
INFO: TransferData.skip_source_errors = False
INFO: TransferData.fail_on_quota_errors = True
INFO: TransferData.delete_destination_extra = False
INFO: TransferData.notify_on_succeeded = True
INFO: TransferData.notify_on_failed = True
INFO: TransferData.notify_on_inactive = True
INFO: TransferClient.get_task(867de1e9-41c0-11f0-b34a-0256b1e82d1f, ...)
INFO: request done (success)
INFO: 20250605_035241_351599: Previous task_id 867de1e9-41c0-11f0-b34a-0256b1e82d1f Still Active. Returning ACTIVE.
INFO: 20250605_035241_351678: SURFACE hpss globus_transfer(name=index.db) returns ACTIVE
DEBUG: 20250605_035241_351756: calling globus_finalize()
INFO: 20250605_035241_351807: FINAL TransferData: accumulated items:
    (finalize) PUSHING (2) source item: /lcrc/group/e3sm2/ac.qtang/tst/zstash/index.db
INFO: 20250605_035241_351862: DIVING: Submit Transfer for tst index
INFO: TransferClient.submit_transfer(...)
INFO: request done (success)
```

### What machine were you running on?

Chrysalis

### Environment

e3sm_unified_1.11.1

### Minimal Complete Verifiable Example (MCVE)

```python

```

### Relevant log output

```python

```

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by forsyth2**
## Summary

Objectives:
- Delete tars _eventually_, even if `--non-blocking` is set

Issue resolution:
- Closes #374 
- Replaces #383 (That one had many conflicts when I rebased on `upstream/main`, so I decided to open a distinct PR to compare the two).

Select one: This pull request is...
- [x] a bug fix: increment the patch version
- [ ] a small improvement: increment the minor version
- [ ] a new feature: increment the minor version
- [ ] an incompatible (non-backwards compatible) API change: increment the major version

## Big Change

- [ ] To merge, I will use "Create a merge commit". That is, this change is large enough to require multiple units of work (i.e., it should be multiple commits).

### 1. Does this do what we want it to do?

Required:
- [x] Product Management: I have confirmed with the stakeholders that the objectives above are correct and complete.
- [x] Testing: I have added at least one automated test. Every objective above is represented in at least one test.
  - Done in #404
- [ ] Testing: I have considered likely and/or severe edge cases and have included them in testing.

If applicable:
- [ ] Testing: this pull request adds at least one new possible command line option. I have tested using this option with and without any other option that may interact with it.
  - N/A

### 2. Are the implementation details accurate & efficient?

Required:
- [x] Logic: I have visually inspected the entire pull request myself.
- [ ] Logic: I have left GitHub comments highlighting important pieces of code logic. I have had these code blocks reviewed by at least one other team member.

If applicable:
- [ ] Dependencies: This pull request introduces a new dependency. I have discussed this requirement with at least one other team member. The dependency is noted in `zstash/conda`, not just an `import` statement.
  - N/A

### 3. Is this well documented?

Required:
- [ ] Documentation: by looking at the docs, a new user could easily understand the functionality introduced by this pull request.
  - This fix makes `zstash` behaves as should be expected based on the docs.

### 4. Is this code clean?

Required:
- [x] Readability: The code is as simple as possible and well-commented, such that a new team member could understand what's happening.
- [x] Pre-commit checks: All the pre-commits checks have passed.

If applicable:
- [ ] Software architecture: I have discussed relevant trade-offs in design decisions with at least one other team member. It is unlikely that this pull request will increase tech debt.

### CODE REVIEW

**Review by forsyth2 — COMMENTED** [review]
Self-review using [Claude's code-review guide](https://github.com/E3SM-Project/zstash/pull/405#issuecomment-3706012340). Currently (as of [commit 4](https://github.com/E3SM-Project/zstash/pull/405/commits/ee98156e44ce56d0309b613b1ee2bf675e00c5b3)), the entire test suite is passing.

<details>

<summary> Self-review </summary>

## Verifications

> Verify: Should is_index check be redundant with proper keep flag usage?

Yes, `index.db` is treated differently than the tar files. `keep` applies to the tar files. The `index.db` is _always_ kept.

> Potential issue: In non-blocking mode, status might still be "ACTIVE" or "UNKNOWN" here
> Check: Is deletion properly deferred to globus_finalize()?

Yes, the test added in #404 appears to confirm this.

> Question: Why delete both curr_transfers and prev_transfers? Is this double-deletion safe?
> Verify the state machine: curr_transfers → prev_transfers → deletion

I believe this is fine. By this point in the code, we've passed the `# Wait for any submitted transfers to complete` block and are just trying to clear the transfer queues.

> Verify: Does this maintain the same behavior?

I believe so; no symlink test was broken.

> Tar files deleted after successful Globus transfer (without --keep)
> Tar files preserved with --keep flag
> index.db never deleted
> Non-blocking mode: files deleted after finalization

Yes, in test added in #404.

> Direct HPSS transfers still work correctly

Yes, no HPSS test was broken.

> Very long transfers with EXHAUSTED_TIMEOUT_RETRIES

This is more the domain of #407 (which was recently determined to already be resolved on `main`).

> --follow-symlinks with broken symlink behavior unchanged

Yes, no symlink test broke.

> --dry-run doesn't trigger deletions

Verified by visual inspection of `--dry-run` code path.

> Database corruption handling unchanged

Yes, that test didn't break.

> In hpss_transfer(), status might be "UNKNOWN" or "ACTIVE" when deletion logic runs. > Files might not get queued for deletion properly.
> Check: Does globus_finalize() compensate for this?

Yes. By this point in the code, we've passed the `# Wait for any submitted transfers to complete` block and are just trying to clear the transfer queues.

> In globus_finalize(), if there's pending transfer_data, it submits a new transfer. But globus_transfer() already submitted data.
> Verify: Is this intentional for batching? Or could it duplicate transfers?

This is explicitly in a `# Check if there's any pending transfer data that hasn't been submitted yet` block.

> When does curr become prev?
> Why delete both in finalize?
> What happens on consecutive calls?

`curr` is the list of tars being transferred via Globus right now. `prev` was the previous list. `prev` is the list of tars that are now ok to delete because they've been successfully transferred. They are both deleted in `finalize` because we have completed `globus_wait` by then and all queues should be cleared.

> If globus_finalize() calls sys.exit(1) (line 236), are files left behind?
> Check: Should there be cleanup in error paths?

We probably should _not_ be deleting in the case of error.

> Lines commented with # Don't track index.db for deletion - verify this is consistently applied everywhere index.db is transferred.

Yes, otherwise the tests would not be passing.

> Why is index.db excluded from htc tracking in some calls but the parameter is commented out?

This is referring to: 
```python
    hpss_put(
        hpss,
        get_db_filename(cache),
        cache,
        keep=args.keep,
        is_index=True,
        gtc=gtc,
        # htc=htc,  # Don't track index.db for deletion
    )
```
It's commented out there to make specific note that we don't need to track `index.db`, only the tars. 

> In non-blocking mode, how do you ensure files are deleted if status is still "ACTIVE" when hpss_transfer() returns?

That's the point of the `wait` logic. 

> Should globus_wait() be called for ALL submitted transfers, not just the most recent?

No, that would render `--non-blocking` useless. The point is to wait on blocking transfers and at the very end. 

> What's the expected behavior if a user Ctrl-C's during finalization?

I imagine we'd want to keep the files since we're not sure they're transferred yet then.

## Not explicitly tested/verified

> Blocking mode: files deleted immediately after each transfer

The test from #404 does test that the files are deleted, but I suppose it's not explicitly testing it's done after each transfer.

> Multiple tar files in single run handled correctly

The test from #404 is only testing one tar file.

> Transfer fails: files should NOT be deleted
> Mixed successful/failed transfers
> Globus endpoint activation failures

How do we force a transfer failure?

>  File already deleted (verify graceful handling)

We'd have to concurrently delete files while the test was transferring data.

## Possible improvements

> Magic values: "SUCCEEDED", "ACTIVE" should be constants

I can make an `enum` for this.

> Consider: Replace List with deque for transfer collections (performance)

That seems reasonable. (There probably aren't _that_ many tars batched together for each transfer, but there's not really a downside to switching to `deque`).

## Possible concerns:
> Thread safety: Is this code thread-safe now? (Likely wasn't before either)

</details>

File: zstash/create.py [hunk 1]
```diff
@@ -52,12 +53,13 @@ def create():
...
 
+    gtc: Optional[GlobusTransferCollection] = None
     if hpss != "none":
         url = urlparse(hpss)
         if url.scheme == "globus":
             # identify globus endpoints
-            logger.debug(f"{ts_utc()}:Calling globus_activate(hpss)")
-            globus_activate(hpss)
+            logger.debug(f"{ts_utc()}:Calling globus_activate()")
+            gtc = globus_activate(hpss)
```
Comment (forsyth2) [inline zstash/create.py]: Using a new object to hold state data, rather than using global variables.

File: zstash/create.py [hunk 2]
```diff
@@ -88,14 +90,23 @@ def create():
...
     logger.debug(f"{ts_utc()}: calling hpss_put() for {get_db_filename(cache)}")
-    hpss_put(hpss, get_db_filename(cache), cache, keep=args.keep, is_index=True)
+    hpss_put(
+        hpss,
+        get_db_filename(cache),
+        cache,
+        keep=args.keep,
+        is_index=True,
+        gtc=gtc,
+        # htc=htc,  # Don't track index.db for deletion
```
Comment (forsyth2) [inline zstash/create.py]: We don't need to track `index.db` because we never delete it.

File: zstash/globus.py [hunk 3]
```diff
@@ -3,162 +3,154 @@
...
-        r = transfer_client.endpoint_autoactivate(ep_id, if_expires_in=600)
+        r = gtc.transfer_client.endpoint_autoactivate(ep_id, if_expires_in=600)
         if r.get("code") == "AutoActivationFailed":
             logger.error(
                 f"The {ep_id} endpoint is not activated or the current activation expires soon. Please go to https://app.globus.org/file-manager/collections/{ep_id} and (re)activate the endpoint."
             )
             sys.exit(1)
-
-
-def file_exists(name: str) -> bool:
```
Comment (forsyth2) [inline zstash/globus.py]: Moved to `zstash/globus_utils.py`

File: zstash/globus.py [hunk 4]
```diff
@@ -174,151 +166,96 @@ def globus_transfer(  # noqa: C901
...
-    if not non_blocking:
-        if task_id:
-            globus_wait(task_id)
-        if last_task_id:
-            globus_wait(last_task_id)
+        # Clean up tar files that were queued for deletion
+        if htc.curr_transfers:
+            delete_transferred_files(htc)
+        if htc.prev_transfers:
+            delete_transferred_files(htc)
```
Comment (forsyth2) [inline zstash/globus.py]: This part is added

File: zstash/hpss.py [hunk 5]
```diff
@@ -129,20 +141,15 @@ def hpss_transfer(
...
 
         if transfer_type == "put":
-            if not keep:
-                if (scheme != "globus") or (globus_status == "SUCCEEDED"):
+            if (not keep) and (not is_index):
+                if scheme != "globus":
+                    # For direct HPSS, delete immediately since transfer is synchronous
+                    # Delete the file that was just transferred
+                    delete_current_files(htc)
+                elif globus_status == "SUCCEEDED":
```
Comment (forsyth2) [inline zstash/hpss.py]: This logic is changed; the HPSS case needs to be treated differently than the Globus case.

File: zstash/transfer_tracking.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+
+# Globus specific #############################################################
+class GlobusTransfer(object):
+    def __init__(self):
+        self.transfer_data: Optional[TransferData] = None
+        self.task_id: Optional[str] = None
+        # https://docs.globus.org/api/transfer/task/#task_fields
+        # ACTIVE, SUCCEEDED, FAILED, INACTIVE
+        self.task_status: Optional[str] = None
```
Comment (forsyth2) [inline zstash/transfer_tracking.py]: Possible improvement: could replace with an `enum`

File: zstash/transfer_tracking.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+    def get_most_recent_transfer(self) -> Optional[GlobusTransfer]:
+        return self.transfers[-1] if self.transfers else None
+
+
+# All Transfers ###############################################################
+class HPSSTransferCollection(object):
+    def __init__(self):
+        self.prev_transfers: List[str] = []  # Can remove
+        self.curr_transfers: List[str] = []  # Still using!
```
Comment (forsyth2) [inline zstash/transfer_tracking.py]: Possible improvement: use `dequeue` rather than `List`

File: zstash/transfer_tracking.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        logger.debug(f"{ts_utc()}: HPSSTransferCollection initialized")
+
+
+def delete_transferred_files(htc: HPSSTransferCollection):
```
Comment (forsyth2) [inline zstash/transfer_tracking.py]: Used in Globus case: we can delete the previous files and rotate the current transfer list to be the "previous" list.

**Review by forsyth2 — COMMENTED** [review]
Adding explanatory notes
